category_encoders/target_encoder.py

```diff
@@ -104,22 +105,33 @@ def __init__(self, verbose=0, cols=None, drop_invariant=False, return_df=True, h
         self.mapping = None
         self._mean = None
         if hierarchy:
-            if self._depth(hierarchy) > 2:
-                raise ValueError('Hierarchy mapping contains too many levels')
-        self.hierarchy = hierarchy
+            self.hierarchy = {}
+            self.hierarchy_depth = {}
+            for switch in hierarchy:
+                D = fd.flatten(hierarchy[switch], inverse=True)
+                hierarchy_check = self._check_dict_key_tuples(D)
+                self.hierarchy_depth[switch] = hierarchy_check[1]
+                if not hierarchy_check[0]:
+                    raise ValueError('Hierarchy mapping contains different levels for key "' + switch + '"')
+                self.hierarchy[switch] = {(k if type(t) is tuple else t): v for t, v in D.items() for k in t}
+        else:
+            self.hierarchy = hierarchy
         self.cols_hier = []
 
-    def _depth(self, d):
-        return max(self._depth(v) if isinstance(v, dict) else 0 for v in d.values()) + 1
+    def _check_dict_key_tuples(self, d):
+        min_tuple_size = min(len(v) for v in d.values())
+        max_tuple_size = max(len(v) for v in d.values())
+        return True if min_tuple_size == max_tuple_size else False, min_tuple_size
```

Comment:
> also in the `True` case you want to return the size right? otherwise `hierarchy_check[1]` might be undefined (c.f. line 113)

This actually does the job despite what it seems at first glance.  The if/else come first then the min_size is added to either afterwards.  So it still returns the min_size in both cases.

Reply:
But I agree it's not easily readable (even for me second time round) so will change it.

Reply:
aaah that's how it works. Thanks for changing anyway